From 7c62573da6cd26564e0a25a2e504f2832f19a6c1 Mon Sep 17 00:00:00 2001
From: Aihua Xu <axu@cloudera.com>
Date: Tue, 26 Apr 2016 19:31:40 -0400
Subject: [PATCH 0584/1164] CDH-38582 : Backport HIVE-13585: Add counter
 metric for direct sql failures (Mohit Sabharwal,
 reviewed by Aihua Xu, Sergey Shelukhin)

(cherry picked from commit 815499af9543687948b0330c1f8793bfbf2dea67)

Change-Id: Id4778567e4434f6863d68002d7514d2987e1b8cf
---
 .../common/metrics/common/MetricsConstant.java     |    2 +
 metastore/pom.xml                                  |    8 ++++
 .../apache/hadoop/hive/metastore/ObjectStore.java  |   16 ++++++-
 .../hadoop/hive/metastore/TestObjectStore.java     |   50 ++++++++++++++++++++
 4 files changed, 74 insertions(+), 2 deletions(-)

diff --git a/common/src/java/org/apache/hadoop/hive/common/metrics/common/MetricsConstant.java b/common/src/java/org/apache/hadoop/hive/common/metrics/common/MetricsConstant.java
index 76ab242..9a5c605 100644
--- a/common/src/java/org/apache/hadoop/hive/common/metrics/common/MetricsConstant.java
+++ b/common/src/java/org/apache/hadoop/hive/common/metrics/common/MetricsConstant.java
@@ -56,4 +56,6 @@
   public static final String DELETE_TOTAL_DATABASES = "delete_total_count_dbs";
   public static final String DELETE_TOTAL_TABLES = "delete_total_count_tables";
   public static final String DELETE_TOTAL_PARTITIONS = "delete_total_count_partitions";
+
+  public static final String DIRECTSQL_ERRORS = "directsql_errors";
 }
diff --git a/metastore/pom.xml b/metastore/pom.xml
index 2dd48b3..135b41a 100644
--- a/metastore/pom.xml
+++ b/metastore/pom.xml
@@ -121,6 +121,14 @@
       <artifactId>libthrift</artifactId>
       <version>${libthrift.version}</version>
     </dependency>
+    <!-- test intra-project -->
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-common</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
+      <scope>test</scope>
+    </dependency>
     <!-- test inter-project -->
     <dependency>
       <groupId>junit</groupId>
diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java b/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
index ddbff0e..b4a9f96 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
@@ -2539,7 +2539,8 @@ private void dropPartitionsNoTxn(String dbName, String tblName, List<String> par
   }
 
   /** Helper class for getting stuff w/transaction, direct SQL, perf logging, etc. */
-  private abstract class GetHelper<T> {
+  @VisibleForTesting
+  public abstract class GetHelper<T> {
     private final boolean isInTxn, doTrace, allowJdo;
     private boolean doUseDirectSql;
     private long start;
@@ -2637,6 +2638,16 @@ private void handleDirectSqlError(Exception ex) throws MetaException, NoSuchObje
       } else {
         start = doTrace ? System.nanoTime() : 0;
       }
+
+      Metrics metrics = MetricsFactory.getInstance();
+      if (metrics != null) {
+        try {
+          metrics.incrementCounter(MetricsConstant.DIRECTSQL_ERRORS);
+        } catch (Exception e) {
+          LOG.warn("Error reporting Direct SQL errors to metrics system", e);
+        }
+      }
+
       doUseDirectSql = false;
     }
 
@@ -2676,7 +2687,8 @@ protected String describeResult() {
     }
   }
 
-  private abstract class GetDbHelper extends GetHelper<Database> {
+  @VisibleForTesting
+  public abstract class GetDbHelper extends GetHelper<Database> {
     /**
      * GetHelper for returning db info using directSql/JDO.
      * Since this is a db-level call, tblName is ignored, and null is passed irrespective of what is passed in.
diff --git a/metastore/src/test/org/apache/hadoop/hive/metastore/TestObjectStore.java b/metastore/src/test/org/apache/hadoop/hive/metastore/TestObjectStore.java
index 71eafd4..d8dbd0f 100644
--- a/metastore/src/test/org/apache/hadoop/hive/metastore/TestObjectStore.java
+++ b/metastore/src/test/org/apache/hadoop/hive/metastore/TestObjectStore.java
@@ -21,6 +21,11 @@
 import java.util.HashMap;
 import java.util.List;
 
+import org.apache.hadoop.hive.common.metrics.common.MetricsConstant;
+import org.apache.hadoop.hive.common.metrics.common.MetricsFactory;
+import org.apache.hadoop.hive.common.metrics.metrics2.CodahaleMetrics;
+import org.apache.hadoop.hive.common.metrics.metrics2.MetricsReporting;
+import org.apache.hadoop.hive.common.metrics.MetricsTestUtils;
 import org.apache.hadoop.hive.conf.HiveConf;
 import org.apache.hadoop.hive.metastore.api.Database;
 import org.apache.hadoop.hive.metastore.api.FieldSchema;
@@ -221,6 +226,51 @@ public void testRoleOps() throws InvalidObjectException, MetaException, NoSuchOb
     objectStore.removeRole(ROLE1);
   }
 
+  @Test
+  public void testDirectSqlErrorMetrics() throws Exception {
+    HiveConf conf = new HiveConf();
+    conf.setBoolVar(HiveConf.ConfVars.HIVE_SERVER2_METRICS_ENABLED, true);
+    conf.setVar(HiveConf.ConfVars.HIVE_METRICS_REPORTER, MetricsReporting.JSON_FILE.name()
+        + "," + MetricsReporting.JMX.name());
+
+    MetricsFactory.init(conf);
+    CodahaleMetrics metrics = (CodahaleMetrics) MetricsFactory.getInstance();
+
+    objectStore.new GetDbHelper("foo", null, true, true) {
+      @Override
+      protected Database getSqlResult(ObjectStore.GetHelper<Database> ctx) throws MetaException {
+        return null;
+      }
+
+      @Override
+      protected Database getJdoResult(ObjectStore.GetHelper<Database> ctx) throws MetaException,
+          NoSuchObjectException {
+        return null;
+      }
+    }.run(false);
+
+    String json = metrics.dumpJson();
+    MetricsTestUtils.verifyMetricsJson(json, MetricsTestUtils.COUNTER,
+        MetricsConstant.DIRECTSQL_ERRORS, "");
+
+    objectStore.new GetDbHelper("foo", null, true, true) {
+      @Override
+      protected Database getSqlResult(ObjectStore.GetHelper<Database> ctx) throws MetaException {
+        throw new RuntimeException();
+      }
+
+      @Override
+      protected Database getJdoResult(ObjectStore.GetHelper<Database> ctx) throws MetaException,
+          NoSuchObjectException {
+        return null;
+      }
+    }.run(false);
+
+    json = metrics.dumpJson();
+    MetricsTestUtils.verifyMetricsJson(json, MetricsTestUtils.COUNTER,
+        MetricsConstant.DIRECTSQL_ERRORS, 1);
+  }
+
   public static void dropAllStoreObjects(RawStore store) throws MetaException, InvalidObjectException, InvalidInputException {
     try {
       Deadline.registerIfNot(100000);
-- 
1.7.9.5

